service/history/transferQueueTaskExecutorBase.go

```diff
@@ -261,7 +261,17 @@ func (t *transferQueueTaskExecutorBase) deleteExecution(ctx context.Context, tas
 		return err
 	}
 
+	// Here, we ensure that the workflow is closed successfully before deleting it. Otherwise, the mutable state
+	// might be deleted before the close task is executed, and so the close task will be dropped. In passive cluster,
+	// this check can be ignored.
+	//
+	// Additionally, this function itself could be called from within the close execution task, so we need to skip
+	// the check in that case because the close execution task would be waiting for itself to finish forever. So, the
+	// ensureNoPendingCloseTask flag is set iff we're running in the active cluster, and we aren't processing the
+	// CloseExecutionTask from within this same goroutine.
```

Comment:
```suggestion
	// Here, we ensure that the delete workflow task is executed after close workflow task. Otherwise, the mutable state
	// will be deleted before the close task is executed, and so the close task will be dropped, and logic associated with workflow closure (notify children and parent workflows, for example) won't be executed. In passive cluster,
	// this check can be ignored because it is ok to drop close task there (logic is executed in active cluster only).
	//
	// Additionally, this function itself could be called from within the close execution task, so we need to skip
	// the check in that case to prevent deadlock (the close execution task would be waiting for itself to finish). So, the
	// ensureNoPendingCloseTask flag is set iff we're running in the active cluster, and we aren't processing the
	// CloseExecutionTask.
```